src/cargo/core/global_cache_tracker.rs

```diff
@@ -0,0 +1,1754 @@
+            Some(i) => Ok(*i),
+            None => {
+                let Some(id) = GlobalCacheTracker::id_from_name(
+                    conn,
+                    REGISTRY_INDEX_TABLE,
+                    &encoded_registry_name,
+                )?
+                else {
+                    bail!("expected registry_index {encoded_registry_name} to exist, but wasn't found");
+                };
+                self.registry_keys.insert(encoded_registry_name, id);
+                Ok(id)
+            }
+        }
+    }
+
+    /// Returns the numeric ID of the git db, either fetching from the local
+    /// cache, or getting it from the database.
+    ///
+    /// It is an error if the git db does not exist.
+    fn git_id(&mut self, conn: &Connection, encoded_git_name: InternedString) -> CargoResult<i64> {
+        match self.git_keys.get(&encoded_git_name) {
+            Some(i) => Ok(*i),
+            None => {
+                let Some(id) =
+                    GlobalCacheTracker::id_from_name(conn, GIT_DB_TABLE, &encoded_git_name)?
+                else {
+                    bail!("expected git_db {encoded_git_name} to exist, but wasn't found")
+                };
+                self.git_keys.insert(encoded_git_name, id);
+                Ok(id)
+            }
+        }
+    }
+}
+
+/// Converts a [`SystemTime`] to a [`Timestamp`] which can be stored in the database.
+fn to_timestamp(t: &SystemTime) -> Timestamp {
+    t.duration_since(SystemTime::UNIX_EPOCH)
+        .expect("invalid clock")
+        .as_secs()
+}
+
+/// Returns the current time.
+///
+/// This supports pretending that the time is different for testing using an
+/// environment variable.
+///
+/// If possible, try to avoid calling this too often since accessing clocks
+/// can be a little slow on some systems.
+#[allow(clippy::disallowed_methods)]
+fn now() -> Timestamp {
+    match std::env::var("__CARGO_TEST_LAST_USE_NOW") {
+        Ok(now) => now.parse().unwrap(),
+        Err(_) => to_timestamp(&SystemTime::now()),
+    }
+}
+
+/// Returns whether or not the given error should cause a warning to be
+/// displayed to the user.
+///
+/// In some situations, like a read-only global cache, we don't want to spam
+/// the user with a warning. I think once cargo has controllable lints, I
+/// think we should consider changing this to always warn, but give the user
+/// an option to silence the warning.
+pub fn is_silent_error(e: &anyhow::Error) -> bool {
+    if let Some(e) = e.downcast_ref::<rusqlite::Error>() {
+        if matches!(
+            e.sqlite_error_code(),
+            Some(ErrorCode::CannotOpen | ErrorCode::ReadOnly)
+        ) {
+            return true;
+        }
+    }
+    false
+}
+
+fn du(path: &Path, table_name: &str) -> CargoResult<u64> {
+    // !.git is used because clones typically use hardlinks for the git
+    // contents. TODO: Verify behavior on Windows.
+    // TODO: Or even better, switch to worktrees, and remove this.
```

Comment:
Whats the plan for tracking and resolving these TODOs?

Reply:
I can add resolving those to the tracking issue.